application/src/main/java/org/togetherjava/tjbot/commands/filesharing/FileSharingMessageListener.java

```
@@ -0,0 +1,198 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+
+/**
+ * Listener that receives all sent help messages and uploads them to a share service if the message
+ * contains a file with the given extension in the
+ * {@link FileSharingMessageListener#extensionFilter}.
+ */
+public class FileSharingMessageListener extends MessageReceiverAdapter {
+
+    private static final ObjectMapper JSON = new ObjectMapper();
+
+    private static final String SHARE_API = "https://api.github.com/gists";
+    private static final HttpClient CLIENT = HttpClient.newHttpClient();
+
+    private final String gistApiKey;
+    private final Set<String> extensionFilter = Set.of("txt", "java", "gradle", "xml", "kt", "json",
+            "fxml", "css", "c", "h", "cpp", "py", "yml");
+
+    private final Predicate<String> isStagingChannelName;
+    private final Predicate<String> isOverviewChannelName;
+
+
+    public FileSharingMessageListener(@NotNull Config config) {
+        super(Pattern.compile(".*"));
+
+        gistApiKey = config.getGistApiKey();
+        isStagingChannelName = Pattern.compile(config.getHelpSystem().getStagingChannelPattern())
+            .asMatchPredicate();
+        isOverviewChannelName = Pattern.compile(config.getHelpSystem().getOverviewChannelPattern())
+            .asMatchPredicate();
+    }
+
+    @Override
+    public void onMessageReceived(@NotNull MessageReceivedEvent event) {
+        User author = event.getAuthor();
+        if (author.isBot() || event.isWebhookMessage()) {
+            return;
+        }
+
+        if (!isHelpThread(event)) {
+            return;
+        }
+
+
+        List<Message.Attachment> attachments = event.getMessage()
+            .getAttachments()
+            .stream()
+            .filter(this::isAttachmentRelevant)
+            .toList();
+
+        CompletableFuture.runAsync(() -> processAttachments(event, attachments));
```

This suppresses failures, unfortunately.

Gonna add a:
```java
runAsync(() -> {
  try {
    processAttachments(event, attachmenets);
  } catch (Exception e) {
    logger.error("Unknown error while processing attachments", e);
  }
}
```